src/core/public/application/application_service.tsx

```diff
@@ -191,17 +199,65 @@ export class ApplicationService {
       registerMountContext: this.mountContext.registerContext,
       getUrlForApp: (appId, { path }: { path?: string } = {}) =>
         getAppUrl(availableMounters, appId, path),
-      navigateToApp: (appId, { path, state }: { path?: string; state?: any } = {}) => {
-        this.navigate!(getAppUrl(availableMounters, appId, path), state);
-        this.currentAppId$.next(appId);
+      navigateToApp: async (appId, { path, state }: { path?: string; state?: any } = {}) => {
+        if (await this.shouldNavigateTo(appId, overlays)) {
+          this.appLeaveHandlers.delete(this.currentAppId$.value!);
+          this.navigate!(getAppUrl(availableMounters, appId, path), state);
+          this.currentAppId$.next(appId);
+        }
+      },
```

Comment:
I tried to unit or integration test this, however after a lot of attempts, I think this is impossible. 

As the `appLeaveHandler` from the currently mounted application is registered when the app is actually mounted by the router, it requires a real end-to-end environment, as mocking the history as it's done in `src/core/public/application/integration_tests/router.test.tsx` is not enough because we want the react router to actually performs the mounting to let the application register it's `AppLeaveHandler` during mount.

I guess the only option is using the FTR with a test plugin registering some applications and navigating between them?